engine/apps/api/permissions/__init__.py

```diff
@@ -1,5 +1,196 @@
+
+        SCHEDULES_READ = _generate_permission_string(Resources.SCHEDULES, Actions.READ)
+        SCHEDULES_WRITE = _generate_permission_string(Resources.SCHEDULES, Actions.WRITE)
+
+        CHATOPS_READ = _generate_permission_string(Resources.CHATOPS, Actions.READ)
+        CHATOPS_WRITE = _generate_permission_string(Resources.CHATOPS, Actions.WRITE)
+
+        OUTGOING_WEBHOOKS_READ = _generate_permission_string(Resources.OUTGOING_WEBHOOKS, Actions.READ)
+        OUTGOING_WEBHOOKS_WRITE = _generate_permission_string(Resources.OUTGOING_WEBHOOKS, Actions.WRITE)
+
+        MAINTENANCE_READ = _generate_permission_string(Resources.MAINTENANCE, Actions.READ)
+        MAINTENANCE_WRITE = _generate_permission_string(Resources.MAINTENANCE, Actions.WRITE)
+
+        API_KEYS_READ = _generate_permission_string(Resources.API_KEYS, Actions.READ)
+        API_KEYS_WRITE = _generate_permission_string(Resources.API_KEYS, Actions.WRITE)
+
+        NOTIFICATION_SETTINGS_READ = _generate_permission_string(Resources.NOTIFICATION_SETTINGS, Actions.READ)
+        NOTIFICATION_SETTINGS_WRITE = _generate_permission_string(Resources.NOTIFICATION_SETTINGS, Actions.WRITE)
+
+        USER_SETTINGS_READ = _generate_permission_string(Resources.USER_SETTINGS, Actions.READ)
+        USER_SETTINGS_WRITE = _generate_permission_string(Resources.USER_SETTINGS, Actions.WRITE)
+        USER_SETTINGS_ADMIN = _generate_permission_string(Resources.USER_SETTINGS, Actions.ADMIN)
+
+        OTHER_SETTINGS_READ = _generate_permission_string(Resources.OTHER_SETTINGS, Actions.READ)
+        OTHER_SETTINGS_WRITE = _generate_permission_string(Resources.OTHER_SETTINGS, Actions.WRITE)
+
+        TESTING = _generate_permission_string(Resources._IMPOSSIBLE_TESTING, Actions._IMPOSSIBLE_TESTING)
+
+    def has_permission(self, request: Request, view: ViewSetOrAPIView) -> bool:
+        action = _get_view_action(request, view)
+
+        rbac_permissions: RBACPermissionsAttribute = getattr(view, RBAC_PERMISSIONS_ATTR, None)
+
+        assert (
+            rbac_permissions is not None
+        ), f"Must define a {RBAC_PERMISSIONS_ATTR} dict on the ViewSet that is consuming the RBACPermission class"
+
+        return has_permissions(request.user.permissions, rbac_permissions.get(action, []))
+
+    def has_object_permission(self, request: Request, view: ViewSetOrAPIView, obj: typing.Any) -> bool:
+        rbac_object_permissions: RBACObjectPermissionsAttribute = getattr(view, RBAC_OBJECT_PERMISSIONS_ATTR, None)
+
+        if rbac_object_permissions:
+            action = _get_view_action(request, view)
+
+            for PermissionClass, actions in rbac_object_permissions.items():
+                if action in actions:
+                    return PermissionClass.has_object_permission(request, view, obj)
+            return False
+
+        # has_object_permission is called after has_permission, so return True if in view there is not
```

Comment:
I'm not sure that `has_object_permission` is called after `has_permission` (at least `rest_framework.permissions.BasePermission` doesn't do it). Do we do that in our code?

Reply:
this seems to have been [a comment](https://github.com/grafana/oncall/blob/1b3b0f9e13f2644d3edce0956112097f691c8f9a/engine/apps/api/permissions/actions.py#L16-L27) that I carried over from `ActionPermission.has_object_permission`:

Reply:
it looks like I carried over [this comment](https://github.com/grafana/oncall/blob/1b3b0f9e13f2644d3edce0956112097f691c8f9a/engine/apps/api/permissions/actions.py#L25-L26) from the previous `ActionPermission` subclass

From the `rest_framework` docs it mentions:

> Note: The instance-level has_object_permission method will only be called if the view-level has_permission 
> checks have already passed. Also note that in order for the instance-level checks to run, the view code should 
> explicitly call .check_object_permissions(request, obj). If you are using the generic views then this will be handled 
> for you by default. (Function-based views will need to check object permissions explicitly, raising 
> PermissionDenied on failure.)